File: backend/pkg/model/prometheus.go
// Package model provides data models for Prometheus integration
package model

import (
	"time"

	"github.com/google/uuid"
)

// PrometheusDataSource represents a Prometheus data source configuration
type PrometheusDataSource struct {
	ID        uuid.UUID `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	CreatedAt time.Time `gorm:"autoCreateTime" json:"createdAt"`
	UpdatedAt time.Time `gorm:"autoUpdateTime" json:"updatedAt"`

	UserID     uuid.UUID `gorm:"type:uuid;not null;index:idx_prometheus_user_id" json:"userId"`
	ClusterID  *uuid.UUID `gorm:"type:uuid;index:idx_prometheus_cluster_id" json:"clusterId,omitempty"`
	Name       string    `gorm:"size:255;not null" json:"name"`
	URL        string    `gorm:"size:2048;not null" json:"url"`
	Username   string    `gorm:"size:255" json:"username,omitempty"`
	Password   string    `gorm:"size:255" json:"-"` // Never expose in JSON
	Status     string    `gorm:"size:50;default:DSStatusActive" json:"status"`

	// TLS configuration
	InsecureSkipTLS bool   `gorm:"default:false" json:"insecureSkipTLS"`
	CACert          string `gorm:"type:text" json:"caCert,omitempty"`
	ClientCert      string `gorm:"type:text" json:"clientCert,omitempty"`
	ClientKey       string `gorm:"type:text" json:"-"` // Never expose in JSON

	// Headers (stored as JSON string)
	Headers string `gorm:"type:text" json:"headers,omitempty"`

	// Test results
	LastTestAt     *time.Time `json:"lastTestAt,omitempty"`
	LastTestStatus string     `gorm:"size:50" json:"lastTestStatus,omitempty"`
	LastTestError  string     `gorm:"type:text" json:"lastTestError,omitempty"`

	// Metrics
	QueryCount       int       `gorm:"default:0" json:"queryCount"`
	LastQueriedAt    *time.Time `json:"lastQueriedAt,omitempty"`
	AverageQueryTime float64   `gorm:"default:0" json:"averageQueryTime"`

	// Relationships
	Cluster *K8sCluster `gorm:"foreignKey:ClusterID" json:"cluster,omitempty"`
}

// PrometheusAlertRule represents a Prometheus alerting rule
type PrometheusAlertRule struct {
	ID        uuid.UUID `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	CreatedAt time.Time `gorm:"autoCreateTime" json:"createdAt"`
	UpdatedAt time.Time `gorm:"autoUpdateTime" json:"updatedAt"`

	UserID       uuid.UUID `gorm:"type:uuid;not null;index:idx_prometheus_alert_user_id" json:"userId"`
	DataSourceID uuid.UUID `gorm:"type:uuid;not null;index:idx_prometheus_alert_datasource_id" json:"dataSourceId"`
	ClusterID    *uuid.UUID `gorm:"type:uuid;index:idx_prometheus_alert_cluster_id" json:"clusterId,omitempty"`
	Name         string    `gorm:"size:255;not null" json:"name"`
	Expression   string    `gorm:"type:text;not null" json:"expression"`
	Duration     int       `gorm:"not null;default:300" json:"duration"` // seconds
	Severity     string    `gorm:"size:50;default:AlertSeverityWarning" json:"severity"`
	Summary      string    `gorm:"type:text" json:"summary,omitempty"`
	Description  string    `gorm:"type:text" json:"description,omitempty"`
	Labels       string    `gorm:"type:text" json:"labels,omitempty"` // JSON string
	Annotations  string    `gorm:"type:text" json:"annotations,omitempty"` // JSON string
	Enabled      bool      `gorm:"default:true" json:"enabled"`

	// Sync status
	Synced       bool      `gorm:"default:false" json:"synced"`
	SyncedAt     *time.Time `json:"syncedAt,omitempty"`
	SyncError    string    `gorm:"type:text" json:"syncError,omitempty"`

	// Statistics
	TriggerCount int        `gorm:"default:0" json:"triggerCount"`
	LastTriggeredAt *time.Time `json:"lastTriggeredAt,omitempty"`

	// Relationships
	DataSource *PrometheusDataSource `gorm:"foreignKey:DataSourceID" json:"dataSource,omitempty"`
	Cluster    *K8sCluster          `gorm:"foreignKey:ClusterID" json:"cluster,omitempty"`
}

// PrometheusQuery represents a query execution record
type PrometheusQuery struct {
	ID        uuid.UUID `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	CreatedAt time.Time `gorm:"autoCreateTime" json:"createdAt"`

	UserID       uuid.UUID `gorm:"type:uuid;not null;index:idx_prometheus_query_user_id" json:"userId"`
	DataSourceID uuid.UUID `gorm:"type:uuid;not null;index:idx_prometheus_query_datasource_id" json:"dataSourceId"`
	Query        string    `gorm:"type:text;not null" json:"query"`
	QueryType    string    `gorm:"size:50;not null" json:"queryType"` // instant, range
	StartTime    string    `gorm:"size:100" json:"startTime,omitempty"`
	EndTime      string    `gorm:"size:100" json:"endTime,omitempty"`
	Step         string    `gorm:"size:50" json:"step,omitempty"`

	Duration      int64   `gorm:"default:0" json:"duration"` // milliseconds
	Success       bool    `gorm:"default:true" json:"success"`
	ErrorMessage  string  `gorm:"type:text" json:"errorMessage,omitempty"`
	ResultCount   int     `gorm:"default:0" json:"resultCount"`

	// Relationships
	DataSource *PrometheusDataSource `gorm:"foreignKey:DataSourceID" json:"dataSource,omitempty"`
}

// PrometheusDashboard represents a custom dashboard configuration
type PrometheusDashboard struct {
	ID        uuid.UUID `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	CreatedAt time.Time `gorm:"autoCreateTime" json:"createdAt"`
	UpdatedAt time.Time `gorm:"autoUpdateTime" json:"updatedAt"`

	UserID       uuid.UUID `gorm:"type:uuid;not null;index:idx_prometheus_dashboard_user_id" json:"userId"`
	ClusterID    *uuid.UUID `gorm:"type:uuid;index:idx_prometheus_dashboard_cluster_id" json:"clusterId,omitempty"`
	Name         string    `gorm:"size:255;not null" json:"name"`
	Description  string    `gorm:"type:text" json:"description,omitempty"`
	Tags         string    `gorm:"type:text" json:"tags,omitempty"` // JSON array
	Config       string    `gorm:"type:text;not null" json:"config"` // JSON: panels, layout, etc.
	IsPublic     bool      `gorm:"default:false" json:"isPublic"`
	Starred      bool      `gorm:"default:false" json:"starred"`
	RefreshRate  int       `gorm:"default:30" json:"refreshRate"` // seconds

	// Relationships
	Cluster *K8sCluster `gorm:"foreignKey:ClusterID" json:"cluster,omitempty"`
}

// DataSource status constants
const (
	DSStatusActive   = "active"
	DSStatusInactive = "inactive"
	DSStatusError    = "error"
)

// Alert severity constants
const (
	AlertSeverityCritical = "critical"
	AlertSeverityWarning  = "warning"
	AlertSeverityInfo     = "info"
)

// CreatePrometheusDataSourceRequest represents a request to create a data source
type CreatePrometheusDataSourceRequest struct {
	ClusterID       *uuid.UUID `json:"clusterId,omitempty"`
	Name            string     `json:"name" binding:"required"`
	URL             string     `json:"url" binding:"required"`
	Username        string     `json:"username,omitempty"`
	Password        string     `json:"password,omitempty"`
	InsecureSkipTLS bool       `json:"insecureSkipTLS,omitempty"`
	CACert          string     `json:"caCert,omitempty"`
	ClientCert      string     `json:"clientCert,omitempty"`
	ClientKey       string     `json:"clientKey,omitempty"`
	Headers         string     `json:"headers,omitempty"`
}

// UpdatePrometheusDataSourceRequest represents a request to update a data source
type UpdatePrometheusDataSourceRequest struct {
	Name            *string  `json:"name,omitempty"`
	URL             *string  `json:"url,omitempty"`
	Username        *string  `json:"username,omitempty"`
	Password        *string  `json:"password,omitempty"`
	InsecureSkipTLS *bool    `json:"insecureSkipTLS,omitempty"`
	CACert          *string  `json:"caCert,omitempty"`
	ClientCert      *string  `json:"clientCert,omitempty"`
	ClientKey       *string  `json:"clientKey,omitempty"`
	Headers         *string  `json:"headers,omitempty"`
	Status          *string  `json:"status,omitempty"`
}

// TestPrometheusDataSourceRequest represents a request to test a data source
type TestPrometheusDataSourceRequest struct {
	URL             string `json:"url" binding:"required"`
	Username        string `json:"username,omitempty"`
	Password        string `json:"password,omitempty"`
	InsecureSkipTLS bool   `json:"insecureSkipTLS,omitempty"`
	CACert          string `json:"caCert,omitempty"`
	ClientCert      string `json:"clientCert,omitempty"`
	ClientKey       string `json:"clientKey,omitempty"`
}

// TestPrometheusDataSourceResponse represents the response from testing a data source
type TestPrometheusDataSourceResponse struct {
	Success    bool   `json:"success"`
	Version    string `json:"version,omitempty"`
	Message    string `json:"message"`
	Error      string `json:"error,omitempty"`
	Duration   int64  `json:"duration"` // milliseconds
}

// CreatePrometheusAlertRuleRequest represents a request to create an alert rule
type CreatePrometheusAlertRuleRequest struct {
	DataSourceID uuid.UUID  `json:"dataSourceId" binding:"required"`
	ClusterID    *uuid.UUID `json:"clusterId,omitempty"`
	Name         string     `json:"name" binding:"required"`
	Expression   string     `json:"expression" binding:"required"`
	Duration     int        `json:"duration" binding:"required,min=1"`
	Severity     string     `json:"severity" binding:"required,oneof=critical warning info"`
	Summary      string     `json:"summary,omitempty"`
	Description  string     `json:"description,omitempty"`
	Labels       string     `json:"labels,omitempty"`
	Annotations  string     `json:"annotations,omitempty"`
}

// UpdatePrometheusAlertRuleRequest represents a request to update an alert rule
type UpdatePrometheusAlertRuleRequest struct {
	Name        *string `json:"name,omitempty"`
	Expression  *string `json:"expression,omitempty"`
	Duration    *int    `json:"duration,omitempty"`
	Severity    *string `json:"severity,omitempty"`
	Summary     *string `json:"summary,omitempty"`
	Description *string `json:"description,omitempty"`
	Labels      *string `json:"labels,omitempty"`
	Annotations *string `json:"annotations,omitempty"`
	Enabled     *bool   `json:"enabled,omitempty"`
}

// PrometheusQueryRequest represents a request to query Prometheus
type PrometheusQueryRequest struct {
	Query     string `json:"query" binding:"required"`
	QueryType string `json:"queryType" binding:"required,oneof=instant range"`
	StartTime string `json:"startTime,omitempty"` // RFC3339 or relative (e.g., "1h ago")
	EndTime   string `json:"endTime,omitempty"`   // RFC3339 or "now"
	Step      string `json:"step,omitempty"`      // e.g., "15s", "1m"
}

// PrometheusQueryResponse represents the response from a Prometheus query
type PrometheusQueryResponse struct {
	Status   string        `json:"status"`
	Data     []PrometheusSeries `json:"data,omitempty"`
	Error    string        `json:"error,omitempty"`
	Duration int64         `json:"duration"` // milliseconds
}

// PrometheusSeries represents a single time series
type PrometheusSeries struct {
	Metric map[string]string `json:"metric"`
	Values []PrometheusValue  `json:"values,omitempty"`
	Value  *PrometheusValue   `json:"value,omitempty"`
}

// PrometheusValue represents a single value at a timestamp
type PrometheusValue struct {
	Timestamp float64 `json:"timestamp"`
	Value     string  `json:"value"`
}

// CreatePrometheusDashboardRequest represents a request to create a dashboard
type CreatePrometheusDashboardRequest struct {
	ClusterID   *uuid.UUID `json:"clusterId,omitempty"`
	Name        string     `json:"name" binding:"required"`
	Description string     `json:"description,omitempty"`
	Tags        string     `json:"tags,omitempty"`
	Config      string     `json:"config" binding:"required"`
	IsPublic    bool       `json:"isPublic,omitempty"`
	RefreshRate int        `json:"refreshRate,omitempty"`
}

// UpdatePrometheusDashboardRequest represents a request to update a dashboard
type UpdatePrometheusDashboardRequest struct {
	Name        *string `json:"name,omitempty"`
	Description *string `json:"description,omitempty"`
	Tags        *string `json:"tags,omitempty"`
	Config      *string `json:"config,omitempty"`
	IsPublic    *bool   `json:"isPublic,omitempty"`
	RefreshRate *int    `json:"refreshRate,omitempty"`
	Starred     *bool   `json:"starred,omitempty"`
}

File: frontend/src/api/prometheus.ts
import axios from 'axios'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8080'

export interface PrometheusDataSource {
  id: string
  userId: string
  clusterId?: string
  name: string
  url: string
  username?: string
  status: 'active' | 'inactive' | 'error'
  insecureSkipTLS: boolean
  caCert?: string
  clientCert?: string
  lastTestAt?: string
  lastTestStatus?: string
  lastTestError?: string
  queryCount: number
  lastQueriedAt?: string
  averageQueryTime: number
  createdAt: string
  updatedAt: string
  cluster?: {
    id: string
    name: string
  }
}

export interface CreatePrometheusDataSourceRequest {
  clusterId?: string
  name: string
  url: string
  username?: string
  password?: string
  insecureSkipTLS?: boolean
  caCert?: string
  clientCert?: string
  clientKey?: string
  headers?: string
}

export interface UpdatePrometheusDataSourceRequest {
  name?: string
  url?: string
  username?: string
  password?: string
  insecureSkipTLS?: boolean
  caCert?: string
  clientCert?: string
  clientKey?: string
  headers?: string
  status?: 'active' | 'inactive' | 'error'
}

export interface TestPrometheusDataSourceRequest {
  url: string
  username?: string
  password?: string
  insecureSkipTLS?: boolean
  caCert?: string
  clientCert?: string
  clientKey?: string
}

export interface TestPrometheusDataSourceResponse {
  success: boolean
  version?: string
  message: string
  error?: string
  duration: number
}

export interface PrometheusAlertRule {
  id: string
  userId: string
  dataSourceId: string
  clusterId?: string
  name: string
  expression: string
  duration: number
  severity: 'critical' | 'warning' | 'info'
  summary?: string
  description?: string
  labels?: string
  annotations?: string
  enabled: boolean
  synced: boolean
  syncedAt?: string
  syncError?: string
  triggerCount: number
  lastTriggeredAt?: string
  createdAt: string
  updatedAt: string
  dataSource?: PrometheusDataSource
  cluster?: {
    id: string
    name: string
  }
}

export interface CreatePrometheusAlertRuleRequest {
  dataSourceId: string
  clusterId?: string
  name: string
  expression: string
  duration: number
  severity: 'critical' | 'warning' | 'info'
  summary?: string
  description?: string
  labels?: string
  annotations?: string
}

export interface UpdatePrometheusAlertRuleRequest {
  name?: string
  expression?: string
  duration?: number
  severity?: 'critical' | 'warning' | 'info'
  summary?: string
  description?: string
  labels?: string
  annotations?: string
  enabled?: boolean
}

export interface PrometheusQueryRequest {
  query: string
  queryType: 'instant' | 'range'
  startTime?: string
  endTime?: string
  step?: string
}

export interface PrometheusQueryResponse {
  status: string
  data?: PrometheusSeries[]
  error?: string
  duration: number
}

export interface PrometheusSeries {
  metric: Record<string, string>
  values?: PrometheusValue[]
  value?: PrometheusValue
}

export interface PrometheusValue {
  timestamp: number
  value: string
}

export interface PrometheusDashboard {
  id: string
  userId: string
  clusterId?: string
  name: string
  description?: string
  tags?: string
  config: string
  isPublic: boolean
  starred: boolean
  refreshRate: number
  createdAt: string
  updatedAt: string
  cluster?: {
    id: string
    name: string
  }
}

export interface CreatePrometheusDashboardRequest {
  clusterId?: string
  name: string
  description?: string
  tags?: string
  config: string
  isPublic?: boolean
  refreshRate?: number
}

export interface UpdatePrometheusDashboardRequest {
  name?: string
  description?: string
  tags?: string
  config?: string
  isPublic?: boolean
  refreshRate?: number
  starred?: boolean
}

const prometheusApi = {
  // ============== Data Source Management ==============

  // List all Prometheus data sources
  getDataSources: async (params?: {
    clusterId?: string
    status?: string
    page?: number
    pageSize?: number
  }) => {
    const response = await axios.get<{
      data: PrometheusDataSource[]
      total: number
      page: number
      pageSize: number
      totalPages: number
    }>(`${API_BASE_URL}/api/v1/prometheus/datasources`, {
      headers: {
        Authorization: `Bearer ${localStorage.getItem('token')}`,
      },
      params,
    })
    return response.data
  },

  // Get a specific data source
  getDataSource: async (id: string) => {
    const response = await axios.get<PrometheusDataSource>(
      `${API_BASE_URL}/api/v1/prometheus/datasources/${id}`,
      {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('token')}`,
        },
      }
    )
    return response.data
  },

  // Create a new data source
  createDataSource: async (data: CreatePrometheusDataSourceRequest) => {
    const response = await axios.post<PrometheusDataSource>(
      `${API_BASE_URL}/api/v1/prometheus/datasources`,
      data,
      {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('token')}`,
        },
      }
    )
    return response.data
  },

  // Update a data source
  updateDataSource: async (
    id: string,
    data: UpdatePrometheusDataSourceRequest
  ) => {
    const response = await axios.put<PrometheusDataSource>(
      `${API_BASE_URL}/api/v1/prometheus/datasources/${id}`,
      data,
      {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('token')}`,
        },
      }
    )
    return response.data
  },

  // Delete a data source
  deleteDataSource: async (id: string) => {
    const response = await axios.delete<{ message: string }>(
      `${API_BASE_URL}/api/v1/prometheus/datasources/${id}`,
      {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('token')}`,
        },
      }
    )
    return response.data
  },

  // Test a data source connection
  testDataSource: async (data: TestPrometheusDataSourceRequest) => {
    const response = await axios.post<TestPrometheusDataSourceResponse>(
      `${API_BASE_URL}/api/v1/prometheus/datasources/test`,
      data,
      {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('token')}`,
        },
      }
    )
    return response.data
  },

  // ============== Alert Rule Management ==============

  // List all alert rules
  getAlertRules: async (params?: {
    dataSourceId?: string
    clusterId?: string
    severity?: string
    enabled?: string
    page?: number
    pageSize?: number
  }) => {
    const response = await axios.get<{
      data: PrometheusAlertRule[]
      total: number
      page: number
      pageSize: number
      totalPages: number
    }>(`${API_BASE_URL}/api/v1/prometheus/alert-rules`, {
      headers: {
        Authorization: `Bearer ${localStorage.getItem('token')}`,
      },
      params,
    })
    return response.data
  },

  // Get a specific alert rule
  getAlertRule: async (id: string) => {
    const response = await axios.get<PrometheusAlertRule>(
      `${API_BASE_URL}/api/v1/prometheus/alert-rules/${id}`,
      {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('token')}`,
        },
      }
    )
    return response.data
  },

  // Create a new alert rule
  createAlertRule: async (data: CreatePrometheusAlertRuleRequest) => {
    const response = await axios.post<PrometheusAlertRule>(
      `${API_BASE_URL}/api/v1/prometheus/alert-rules`,
      data,
      {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('token')}`,
        },
      }
    )
    return response.data
  },

  // Update an alert rule
  updateAlertRule: async (
    id: string,
    data: UpdatePrometheusAlertRuleRequest
  ) => {
    const response = await axios.put<PrometheusAlertRule>(
      `${API_BASE_URL}/api/v1/prometheus/alert-rules/${id}`,
      data,
      {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('token')}`,
        },
      }
    )
    return response.data
  },

  // Delete an alert rule
  deleteAlertRule: async (id: string) => {
    const response = await axios.delete<{ message: string }>(
      `${API_BASE_URL}/api/v1/prometheus/alert-rules/${id}`,
      {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('token')}`,
        },
      }
    )
    return response.data
  },

  // ============== Query Execution ==============

  // Execute a Prometheus query
  executeQuery: async (dataSourceId: string, data: PrometheusQueryRequest) => {
    const response = await axios.post<PrometheusQueryResponse>(
      `${API_BASE_URL}/api/v1/prometheus/datasources/${dataSourceId}/query`,
      data,
      {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('token')}`,
        },
      }
    )
    return response.data
  },

  // ============== Dashboard Management ==============

  // List all dashboards
  getDashboards: async (params?: {
    clusterId?: string
    starred?: string
    page?: number
    pageSize?: number
  }) => {
    const response = await axios.get<{
      data: PrometheusDashboard[]
      total: number
      page: number
      pageSize: number
      totalPages: number
    }>(`${API_BASE_URL}/api/v1/prometheus/dashboards`, {
      headers: {
        Authorization: `Bearer ${localStorage.getItem('token')}`,
      },
      params,
    })
    return response.data
  },

  // Get a specific dashboard
  getDashboard: async (id: string) => {
    const response = await axios.get<PrometheusDashboard>(
      `${API_BASE_URL}/api/v1/prometheus/dashboards/${id}`,
      {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('token')}`,
        },
      }
    )
    return response.data
  },

  // Create a new dashboard
  createDashboard: async (data: CreatePrometheusDashboardRequest) => {
    const response = await axios.post<PrometheusDashboard>(
      `${API_BASE_URL}/api/v1/prometheus/dashboards`,
      data,
      {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('token')}`,
        },
      }
    )
    return response.data
  },

  // Update a dashboard
  updateDashboard: async (
    id: string,
    data: UpdatePrometheusDashboardRequest
  ) => {
    const response = await axios.put<PrometheusDashboard>(
      `${API_BASE_URL}/api/v1/prometheus/dashboards/${id}`,
      data,
      {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('token')}`,
        },
      }
    )
    return response.data
  },

  // Delete a dashboard
  deleteDashboard: async (id: string) => {
    const response = await axios.delete<{ message: string }>(
      `${API_BASE_URL}/api/v1/prometheus/dashboards/${id}`,
      {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('token')}`,
        },
      }
    )
    return response.data
  },
}

export default prometheusApi
